### PR TITLE
provider/cloudstack: add project support for CloudStack Instances

### PR DESCRIPTION
**Pull request by jalemieux**
Add the ability to specify the project an instance should belong to. This is contingent on https://github.com/xanzy/go-cloudstack/pull/21/files getting merged.

### CODE REVIEW

**Comment by svanharmelen** [comment]
@jalemieux I just merged PR xanzy/go-cloudstack#21 so we're good to move forward with this one.

I'll do a review in just a bit, but noticed you only added a parameter `project_name` to the `cloudstack_instance` resource. I think if you would really want to work with the CloudStack provider and projects, we should add a (preferably called just `project`) parameter to all resources.

Otherwise when you create a VPC/network and want to create an instance in there, it will not be possible as the VPC/network will not be part of the project. Not sure how CS will handle that, but I would assume that you would want all needed resources to be created within the same project right?

But no need to do everything at once of course, so we can just start with this first resource...

File: builtin/providers/cloudstack/resource_cloudstack_instance.go [hunk 1]
```diff
@@ -82,6 +82,11 @@ func resourceCloudStackInstance() *schema.Resource {
 				Optional: true,
 				Default:  false,
 			},
+			"project_name": &schema.Schema{
```
Comment (svanharmelen) [inline builtin/providers/cloudstack/resource_cloudstack_instance.go]: Just a small nit, please add an empty line between parameter declarations and I would prefer to keep the name a little shorter to just `project`...

File: builtin/providers/cloudstack/resource_cloudstack_instance.go [hunk 2]
```diff
@@ -82,6 +82,11 @@ func resourceCloudStackInstance() *schema.Resource {
...
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  nil,
```
Comment (svanharmelen) [inline builtin/providers/cloudstack/resource_cloudstack_instance.go]: This should be removed as it servers no purpose here. `Default` is only useful when you want to assign an actual default value if the user didn't specify anything. If the user doesn't specify a value the types zero value will be used. And in Go for strings that is and empty string so `""` and not `nil` (in Go a string cannot be `nil`).

Comment (svanharmelen) [inline builtin/providers/cloudstack/resource_cloudstack_instance.go]: Is it possible to move a VM from one project to another? If so the code to do so should be added in the update function of this resource. If not (what I suspect, but I didn't check it) we should add `ForceNew: true` here indicating that when you change the project for this resource, it needs to recreated.

File: builtin/providers/cloudstack/resource_cloudstack_instance.go [hunk 3]
```diff
@@ -148,6 +153,17 @@ func resourceCloudStackInstanceCreate(d *schema.ResourceData, meta interface{})
...
 
+	// If the project_name contains any info, we retreive the project_id
+	if projectName, ok := d.GetOk("project_name"); ok {
+		project, _, err := cs.Project.GetProjectByName(projectName.(string))
+		if err != nil {
+			return err
+		}
+		log.Printf("[DEBUG] project id %s", project.Id)
+		p.SetProjectid(project.Id)
+		d.Set("project_id", project.Id)
```
Comment (svanharmelen) [inline builtin/providers/cloudstack/resource_cloudstack_instance.go]: You can (and should) delete this line. You are trying to set a value for the parameter `project_id`, but that parameter actually doesn't exist (it's not specified in the schema). So even if this doesn't trigger an error (not sure about that), the value will not be saved.

File: builtin/providers/cloudstack/resource_cloudstack_instance.go [hunk 4]
```diff
@@ -148,6 +153,17 @@ func resourceCloudStackInstanceCreate(d *schema.ResourceData, meta interface{})
 		p.SetUserdata(ud)
 	}
 
+	// If the project_name contains any info, we retreive the project_id
+	if projectName, ok := d.GetOk("project_name"); ok {
+		project, _, err := cs.Project.GetProjectByName(projectName.(string))
```
Comment (svanharmelen) [inline builtin/providers/cloudstack/resource_cloudstack_instance.go]: I guess we should extent [this](https://github.com/hashicorp/terraform/blob/master/builtin/providers/cloudstack/resources.go#L30) function so we can use the more generally used pattern:

```go
		projectid, e := retrieveUUID(cs, "project", d.Get("project").(string))
		if e != nil {
			return e.Error()
		}
```

Also this parameter should be a good candidate to accept both a UUID or a name, which will be enabled instantly when using this approach.

**Comment by svanharmelen** [comment]
@jalemieux please see the inline comments for some pointers and suggestions. Also I'm missing a line like `setValueOrUUID(d, "project", vm.Project, vm.Projectid)` in the `resourceCloudStackInstanceRead` function so the value is updated when we do a `terraform refresh`.

So there's still some work to do before we can merge this in, but I really like to direction you are going by adding project support to the CloudStack provider. So thanks for opening the PR and let's make sure we get this feature in!

**Comment by jalemieux** [comment]
@svanharmelen  thanks for the feedback!
I think I have most your suggestions in. I wasn't able to change the project of a VM, so ForceNew make sense. Also Added project support in retrieveUUID. 
Did a few successful round of testing, but notice some weirdness once when removing the project from a resource in my test .tf. I wasn't able to reproduce. 
Maybe you can give me some hints on what happens when the engine realize a previous configuration is missing, couldn't find anything in  helper/schema that seemed helpful for this.

**Comment by svanharmelen** [comment]
@jalemieux nice updates, looks good! Think we should probably add a test for this as well (or extend one of the existing tests).

What happens (or should happen) if TF notices a configuration option is no longer set, is that it will check the new config against your (refreshed) state and if there is a diff it should (when running `apply`) try to update whatever is needed to get in sync again.

So in this case if the instance in member of a project and you remove the project attribute from your config, I think it should notice that and recreate the instance outside a project...

**Comment by jalemieux** [comment]
@svanharmelen thanks, I will work on the tests. 

Regarding the removal of `project` attribute, what you described is in line with my expectation. However, what I see is different. For example, I have `project = "terraform"` in my .tf, for a given `cloustack_instance resource`. If i removed `project = "terraform"` from the .tf, and rerun plan i will get the expected 

    ~ cloudstack_instance.tf4
        project: "terraform" => ""`

However, applying the plan doesn't seem to have any effect. After applying, rerunning plan will produce the same output: 

    ~ cloudstack_instance.tf4
        project: "terraform" => ""`

I figured other providers must run into similar corner cases. There ought to be something that in TF looks for removed attributes and recreate resources if necessary. Or something in the attribute schema itself, that, like `Forcenew`, forces the resource recreation when the attribute is removed (vs changed).

**Comment by svanharmelen** [comment]
Thanks for the clear explanation. I'll have a closer look at this one to see if I can see why this happens.

**Comment by jalemieux** [comment]
@svanharmelen i added a test for create instance with a project's name and project's id.  I tested the former here, but can't test the latter. Worked well with project name.

File: builtin/providers/cloudstack/resource_cloudstack_instance_test.go [hunk 5]
```diff
@@ -233,3 +275,38 @@ resource "cloudstack_instance" "foobar" {
...
+  template = "%s"
+  zone = "%s"
+  expunge = true
+  project = "%s"
+}`,
+	CLOUDSTACK_SERVICE_OFFERING_1,
+	CLOUDSTACK_NETWORK_1,
+	CLOUDSTACK_TEMPLATE,
+	CLOUDSTACK_ZONE,
+	CLOUDSTACK_PROJECT_id)
```
Comment (svanharmelen) [inline builtin/providers/cloudstack/resource_cloudstack_instance_test.go]: Typo breaking Travis...

**Comment by jalemieux** [comment]
@svanharmelen please review. Tested both cases.

**Comment by svanharmelen** [comment]
LGTM!

**Pull request merged**